Code/Framework/AzCore/AzCore/DOM/DomValue.cpp

```diff
@@ -0,0 +1,1305 @@
+        case Type::String:
+            SetString("");
+            break;
+        case Type::Int64:
+            m_value = int64_t{};
+            break;
+        case Type::Uint64:
+            m_value = uint64_t{};
+            break;
+        case Type::Double:
+            m_value = double{};
+            break;
+        case Type::Node:
+            SetNode("");
+            break;
+        case Type::Opaque:
+            AZ_Assert(false, "AZ::Dom::Value may not be constructed with an empty opaque type");
+            break;
+        }
+    }
+
+    Value& Value::operator=(const Value& other)
+    {
+        m_value = other.m_value;
+        return *this;
+    }
+
+    Value& Value::operator=(Value&& other) noexcept
+    {
+        SetNull();
+        memcpy(this, &other, sizeof(Value));
+        memset(&other, 0, sizeof(Value));
+        return *this;
+    }
+
+    bool Value::operator==(const Value& rhs) const
+    {
+        if (IsString() && rhs.IsString())
+        {
+            return GetString() == rhs.GetString();
+        }
+        else if (IsNumber() && rhs.IsNumber())
+        {
+            if (IsInt())
+            {
+                return GetInt64() == rhs.GetInt64();
+            }
+            else if (IsUint())
+            {
+                return GetUint64() == rhs.GetUint64();
+            }
+            else
+            {
+                return GetDouble() == rhs.GetDouble();
+            }
+        }
+        else
+        {
+            return m_value == rhs.m_value;
+        }
+    }
+
+    bool Value::operator!=(const Value& rhs) const
+    {
+        return !operator==(rhs);
+    }
+
+    void Value::Swap(Value& other) noexcept
+    {
+        AZStd::aligned_storage_for_t<Value> temp;
+        memcpy(&temp, this, sizeof(Value));
+        memcpy(this, &other, sizeof(Value));
+        memcpy(&other, &temp, sizeof(Value));
+    }
+
+    Type Dom::Value::GetType() const
+    {
+        switch (m_value.index())
+        {
+        case GetTypeIndex<AZStd::monostate>():
+            return Type::Null;
+        case GetTypeIndex<int64_t>():
+            return Type::Int64;
+        case GetTypeIndex<uint64_t>():
+            return Type::Uint64;
+        case GetTypeIndex<double>():
+            return Type::Double;
+        case GetTypeIndex<bool>():
+            return Type::Bool;
+        case GetTypeIndex<AZStd::string_view>():
+        case GetTypeIndex<SharedStringType>():
+        case GetTypeIndex<ShortStringType>():
+            return Type::String;
+        case GetTypeIndex<ObjectPtr>():
+            return Type::Object;
+        case GetTypeIndex<ArrayPtr>():
+            return Type::Array;
+        case GetTypeIndex<NodePtr>():
+            return Type::Node;
+        case GetTypeIndex<AZStd::shared_ptr<AZStd::any>>():
+            return Type::Opaque;
+        }
+        AZ_Assert(false, "AZ::Dom::Value::GetType: m_value has an unexpected type");
+        return Type::Null;
+    }
```

Comment:
I'll play with it, but fwiw I had a compiler (I think MSVC) issue with a `static_assert` in the type not found in `GetTypeIndex` case that may also apply here.

Reply:
Ah it was clang - same issue with this approach, womp womp:
```
[2022-01-05T19:41:23.480Z] FAILED: Code/Framework/AzCore/CMakeFiles/AzCore.dir/profile/AzCore/DOM/DomValue.cpp.o 

[2022-01-05T19:41:23.480Z] /usr/bin/clang++-12 -DAZ_BUILD_CONFIGURATION_TYPE=\"profile\" -DAZ_ENABLE_DEBUG_TOOLS -DAZ_ENABLE_TRACING -DAZ_PROFILE_BUILD=1 -DLINUX -DLINUX64 -DNDEBUG -D_HAS_EXCEPTIONS=0 -D_PROFILE -D__linux__ -DCMAKE_INTDIR=\"profile\" -I/data/workspace/o3de/Code/Framework/AzCore/. -I/data/workspace/o3de/Code/Framework/AzCore/Platform/Linux -I/data/workspace/o3de/Code/Framework/AzCore/Platform/Common -isystem /data/workspace/3rdParty/packages/unwind-1.2.1-linux/unwind/include -isystem /data/workspace/3rdParty/packages/Lua-5.3.5-rev5-linux/Lua/include -isystem /data/workspace/3rdParty/packages/RapidJSON-1.1.0-rev1-multiplatform/RapidJSON/include -isystem /data/workspace/3rdParty/packages/RapidXML-1.13-rev1-multiplatform/RapidXML/include -isystem /data/workspace/3rdParty/packages/zlib-1.2.11-rev5-linux/zlib/include -isystem /data/workspace/3rdParty/packages/zstd-1.35-multiplatform/zstd/lib -isystem /data/workspace/3rdParty/packages/cityhash-1.1-multiplatform/cityhash/src -fno-exceptions -fvisibility=hidden -fvisibility-inlines-hidden -Wall -Werror -Wno-inconsistent-missing-override -Wrange-loop-analysis -Wno-unknown-warning-option -Wno-parentheses -Wno-reorder -Wno-switch -Wno-undefined-var-template -fPIC -msse4.1  -O2 -g -std=c++17 -MD -MT Code/Framework/AzCore/CMakeFiles/AzCore.dir/profile/AzCore/DOM/DomValue.cpp.o -MF Code/Framework/AzCore/CMakeFiles/AzCore.dir/profile/AzCore/DOM/DomValue.cpp.o.d -o Code/Framework/AzCore/CMakeFiles/AzCore.dir/profile/AzCore/DOM/DomValue.cpp.o -c /data/workspace/o3de/Code/Framework/AzCore/AzCore/DOM/DomValue.cpp

[2022-01-05T19:41:23.480Z] /data/workspace/o3de/Code/Framework/AzCore/AzCore/DOM/DomValue.cpp:355:21: error: static_assert failed "AZ::Dom::Value::GetType: m_value has an unexpected type"

[2022-01-05T19:41:23.480Z]                     static_assert(false, "AZ::Dom::Value::GetType: m_value has an unexpected type");
```

I've just made it a runtime assert for now

Reply:
Surfacing for historic visibility - we went with the `visit` implementation with this PR, but we've found that in-practice `visit` can fail to inline and be a performance pessimization vs reading `variant.index()` directly, so this may be tuned for performance reasons later.